### ISSUE TITLE
Problem with Hibernate PersistentSet

### ISSUE THREAD

**Issue by paschm** [smallrye/smallrye-graphql#95]
I am using JPA with Hibernate implementation to persist my beans. I am using the same beans in the GraphQL-API. In one of these beans I have a OneToMany-mapping to a Set

```
	@OneToMany
	@JoinColumn(name = "konzernId")
	private Set<Gesellschaft> gesellschaften;
```

If I load these beans at runtime this will be a org.hibernate.collection.internal.PersistentSet. When serializing the GraphQL implementation creates a new instance of the PersistentSet in

`io.smallrye.graphql.execution.datafetcher.AnnotatedPropertyDataFetcher.get(AnnotatedPropertyDataFetcher.java:73)`

and copies the elements of the previous PersistentSet in

`io.smallrye.graphql.execution.datafetcher.AnnotatedPropertyDataFetcher.get(AnnotatedPropertyDataFetcher.java:75).`

However this leads to a

```
org.hibernate.LazyInitializationException: failed to lazily initialize a collection, could not initialize proxy - no Session
...
at org.hibernate@5.3.13.Final//org.hibernate.collection.internal.AbstractPersistentCollection.initialize(AbstractPersistentCollection.java:581)
at org.hibernate@5.3.13.Final//org.hibernate.collection.internal.PersistentSet.add(PersistentSet.java:210)
```

as the session is already and correctly closed at this position. This issue is similar to [Hibernate JPA fetch type LAZY #71](https://github.com/smallrye/smallrye-graphql/issues/71), however the Set is correctly fetched when the beans are loaded. I also have a REST-API and there the beans including this Set are correctly serialized. The problem only occurs when calling the add-method at the newly initialized PersistentSet.

If the elemtens of the PersistentSet must be copied to a new Set maybe it would be possible to use a new instance of HashSet instead of PersistentSet? As it seems the used Collection-type is determined in

CollectionHelper.getCorrectCollectionType(Class)

Thank you very much.

**Comment by paschm** [comment]
I am currently using this quick + dirty fix (with no dependency to Hibernate). I replaced getCorrectCollectionType in CollectionHelper by 

```
    public Collection getCorrectCollectionType(Class type) {

        if (type.equals(Collection.class) || type.equals(List.class) || type.getName().equals(ARRAYS_ARRAYLIST)) {
            return new ArrayList();
        } else if (type.equals(Set.class) || "org.hibernate.collection.internal.PersistentSet".equals(type.getName())) {
        	return new HashSet();
        } else {
            try {
                return (Collection) type.newInstance();
            } catch (InstantiationException | IllegalAccessException ex) {
                LOG.error("Can not create new collection of [" + type.getName() + "]");
                return new ArrayList(); // default ?
            }
        }
    }
```

**Comment by phillip-kruger** [comment]
@t1 maybe you want to look at this ?

**Comment by t1** [comment]
#98 should also fix this, I guess.

**Comment by phillip-kruger** [comment]
Hi @paschm .
#98 has been merged. Can you confirm that this is fixed ? 
(You will need to build against the SNAPSHOT - i.e build this master on your PC)

Let me know. Thanks

**Comment by paschm** [comment]
Hello @phillip-kruger ,

unfortunately the fix for #98 doesn't work, see [stacktrace.txt](https://github.com/smallrye/smallrye-graphql/files/4410335/stacktrace.txt).

Same problem as before: A new instance of PersistentSet is created, to which the elements of the previous PersistentSet are added.

The relevant method which creates the new collection instance in CollectionHelper is now

 ```
   public static Collection<?> newCollection(Class<?> type) {
        try {
            return (Collection<?>) type.newInstance();
        } catch (Exception ex) {
            LOG.debug("Cannot create no-arg instance of [" + (type == null ? "null" : type.getName()) + "]", ex);
        }
        if (Set.class.isAssignableFrom(type)) {
            return new HashSet<>();
        }
        return new ArrayList<>();
    }

```
If
```
        if (Set.class.isAssignableFrom(type)) {
            return new HashSet<>();
        }
```
would be the first lines, then it would work. However every Set would then be transformed to a HashSet and I don't think this is intended? (SortedSet, etc.).

The dirty way would be to explicitely handle Hibernate-PersistentCollections, for PersistentSets see my previous  [comment](https://github.com/smallrye/smallrye-graphql/issues/95#issuecomment-606056424).

**Comment by t1** [comment]
@phillip-kruger: if we could get away without cloning collections, things would become much easier. So why is this necessary?

**Comment by phillip-kruger** [comment]
I was also thinking about this. It's used when making a call to a method using reflection. When the method has a collection type parameter.
So we just need the class type (not making a copy of the object).
I think we can change this to not use the type from the runtime, but rather the type from the method signature. I'll look into that.

**Comment by phillip-kruger** [comment]
OK, so I could not get rid of the new Collection functionality. It's used when there are transformation on a collection like return type or collection like parameter. So we basically go through all elements in the collection and transform then and then need to add them to a new collection (of the same type).

I at least changed that this new collection only happen when there is transformation and made sure that the collection type is what is specified in the method signature, and not the underlying implementation of the runtime object.

I am not sure if this works, as we do not have a test case for this. 

@paschm Would you mind checking again against your code (using the latest from master) ? Please let us know here.

If possible, it would be great if you can add a test case to simulate your issue.

Thanks :)

**Comment by paschm** [comment]
Hello @phillip-kruger ,

> @paschm Would you mind checking again against your code (using the latest from master) ? Please let us know here.

The current master is working, thank you very much.

> If possible, it would be great if you can add a test case to simulate your issue.

Sorry, I don't know how to do this. One could use a mocking framework, but I think this won't create PersistentCollections at runtime. So I guess at least you would need a real data source. And this isn't suitable for a unit test...

> It's used when there are transformation on a collection like return type or collection like parameter.

Could you please explain in which cases these transformations occur? In the code I saw 

```
    private boolean shouldTransform() {
        return dateTimeFormatter != null || numberFormat != null;
    }
```

so I guess this transformation occurs, if I hava a collection on which elements a formatting should be applied. Are there other cases?

In general I think it can be problematic to create and use new instances of any arbitrary collection. First it isn't guaranteed that it is instantiable by an no-args constructor. You handled this case accordingly. Second the behaviour of the new instance could be problematic, as seen with the add-method of the Hibernate-PersistentCollections.

Maybe it would be an idea to provide a possibility to supply a custom implementation for the creation of a new collection instance? In 1.0.0 the CollectionHelper is still a CDI managed bean, so one could do something like

```
@Alternative
@Dependent
public class MyCollectionHelper extends CollectionHelper {
   
	public Collection getCorrectCollectionType(Class type) {
    	if("org.hibernate.collection.internal.PersistentSet".equals(type.getName())) {
    		return new HashSet();
    	}
    	return super.getCorrectCollectionType(type);
    }
}
```
However this won't work in the current master, as it is no longer a CDI bean and provides a static method. Or maybe there are other options?

**Comment by phillip-kruger** [comment]
Hi @paschm - yes we are trying to limit the use of CDI. You can check if you get the original error by adding a transformation on the Collection. To do this:

1) the collection needs to be of a type that can be transformed, example LocalDate.
2) add the `@DateFormat` on the collection.

example:
```
public void setImportantDates(List<@DateFormat("dd/MM/yy") LocalDate> importantDates) {
        this.importantDates = importantDates;
}
```

see:
- https://download.eclipse.org/microprofile/microprofile-graphql-1.0.1/microprofile-graphql.html#_dates)
- https://github.com/eclipse/microprofile-graphql/blob/8bedbf687a248c3efa196a0124b4e2df07acf5df/server/tck/src/main/java/org/eclipse/microprofile/graphql/tck/apps/superhero/model/SuperHero.java#L270

Then see if you get the original error. That should be a good test.

Let us know !

**Comment by paschm** [comment]
Hello @phillip-kruger ,

I tried to reproduce the error in the current master with a PersistentSet of LocalDates and a @DateFormat.

I used this code:

```
@Entity
@Table(name = "TEST")
public class Test {

	@Id
	private Integer testId;
	
	@ElementCollection
	@CollectionTable(name = "TESTDATES", joinColumns = @JoinColumn(name = "TESTID"))
	@Column(name = "DATEVALUE")
	private Set<LocalDate> dates;

	public Integer getTestId() {
		return testId;
	}

	public void setTestId(Integer testId) {
		this.testId = testId;
	}

	public Set<@DateFormat("dd.MM.yyyy") LocalDate> getDates() {
		return dates;
	}

	public void setDates(Set<LocalDate> dates) {
		this.dates = dates;
	}
}
```
Script for creating the mapped database schema in orcale db:

```
	create table test (
	    testid integer primary key
	);
	
	create table testdates (
	        testid integer,
	        datevalue date
	);
	
	insert into test
	values (1);
	
	insert into testdates 
	values (1, sysdate);
	
	commit;
```
And now guess what, this works too. The Set of dates is an PersistentSet at runtime, as expected. However for the transformation a new instance of HashSet instead of PersistentSet is returned by the CollectionHelper. The reason is that the Jandex-Type is now used as the parameter for the CollectionHelper. And in the CollectionHelper no instance can be created (see 
[NoSuchMethodException.txt](https://github.com/smallrye/smallrye-graphql/files/4425366/NoSuchMethodException.txt)) and a HashSet is returned as fallback. I also made a 
[screenshot](https://user-images.githubusercontent.com/7092870/78330630-b3bf1a00-7584-11ea-8c55-ae1fef6ec247.png) of my debugging session.

In 1.0.0 I am getting the old [stacktrace](https://github.com/smallrye/smallrye-graphql/files/4425376/stacktrace-1.0.0.txt) as expected.

So everything works fine, although I don't know, if this behaviour is the expected one. For me it is sufficient that collections without a format-annotation don't get transformed any more.
